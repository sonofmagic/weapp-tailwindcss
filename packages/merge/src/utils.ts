import type { ClassValue } from 'clsx'
import type {
  createTailwindMerge as _createTailwindMergeV3,
  extendTailwindMerge as _extendTailwindMergeV3,
  twJoin as _twJoinV3,
  twMerge as _twMergeV3,
} from 'tailwind-merge'
import type {
  createTailwindMerge as _createTailwindMergeV2,
  extendTailwindMerge as _extendTailwindMergeV2,
  twJoin as _twJoinV2,
  twMerge as _twMergeV2,
} from 'tailwind-merge-v2'
import type { CreateOptions } from './types'
import { clsx } from 'clsx'
import { escape } from 'weapp-tailwindcss/escape'

export type TailwindMergeVersion = 2 | 3
// 处理 escape 逻辑
export function noop(x: string) {
  return x
}

export interface CreateFactoryOptions<V extends TailwindMergeVersion> {
  version: V
  twMerge: V extends 3 ? typeof _twMergeV3 : typeof _twMergeV2
  twJoin: V extends 3 ? typeof _twJoinV3 : typeof _twJoinV2
  extendTailwindMerge: V extends 3 ? typeof _extendTailwindMergeV3 : typeof _extendTailwindMergeV2
  createTailwindMerge: V extends 3 ? typeof _createTailwindMergeV3 : typeof _createTailwindMergeV2
}

export function createFactory<V extends TailwindMergeVersion>(opts: CreateFactoryOptions<V>) {
  const {
    twMerge: _twMerge,
    twJoin: _twJoin,
    createTailwindMerge: _createTailwindMerge,
    extendTailwindMerge: _extendTailwindMerge,
  } = opts

  return function create(options?: CreateOptions) {
    const escapeFn = options?.escapeFn ?? escape
    const e = options?.disableEscape ? noop : escapeFn

    function twMerge(...inputs: ClassValue[]): string {
      return e(_twMerge(clsx(...inputs)))
    }

    function twJoin(...inputs: ClassValue[]): string {
      return e(_twJoin(clsx(...inputs)))
    }

    // 版本安全的 extendTailwindMerge
    const extendTailwindMerge = (...args: Parameters<typeof _extendTailwindMerge>) => {
      // @ts-ignore
      const customTwMerge = _extendTailwindMerge(...args)
      return function cn(...inputs: ClassValue[]): string {
        return e(customTwMerge(clsx(...inputs)))
      }
    }

    // 版本安全的 createTailwindMerge
    const createTailwindMerge = (...args: Parameters<typeof _createTailwindMerge>) => {
      // @ts-ignore
      const customTwMerge = _createTailwindMerge(...args)
      return function cn(...inputs: ClassValue[]): string {
        return e(customTwMerge(clsx(...inputs)))
      }
    }

    return {
      twMerge,
      twJoin,
      extendTailwindMerge,
      createTailwindMerge,
    }
  }
}
