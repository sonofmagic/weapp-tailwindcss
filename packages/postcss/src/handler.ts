import type { IStyleHandlerOptions } from './types'
import { defu, defuOverrideArray } from '@weapp-tailwindcss/shared'
import postcss from 'postcss'
import { getDefaultOptions } from './defaults'
import { getPlugins } from './plugins'
import { createInjectPreflight } from './preflight'

function styleHandler(rawSource: string, options: IStyleHandlerOptions) {
  return postcss(
    getPlugins(options),
  )
    .process(
      rawSource,
      defu(options.postcssOptions?.options, {
        from: undefined,
      }),
    )
    .async()
}

export function createStyleHandler(options?: Partial<IStyleHandlerOptions>) {
  const cachedOptions = defuOverrideArray<
    IStyleHandlerOptions,
    Partial<IStyleHandlerOptions>[]
  >(
    options as IStyleHandlerOptions,
    getDefaultOptions(options),
  )

  cachedOptions.cssInjectPreflight = createInjectPreflight(cachedOptions.cssPreflight)

  return (rawSource: string, opt?: Partial<IStyleHandlerOptions>) => {
    return styleHandler(
      rawSource,
      defuOverrideArray<
        IStyleHandlerOptions,
        Partial<IStyleHandlerOptions>[]
      >(opt as IStyleHandlerOptions, cachedOptions),
    )
  }
}
