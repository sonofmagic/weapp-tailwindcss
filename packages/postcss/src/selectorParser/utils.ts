// 选择器 AST 相关的通用工具，提供组合与替换的辅助函数
import type { Node } from 'postcss-selector-parser'
import type { IStyleHandlerOptions } from '../types'
import psp from 'postcss-selector-parser'

export type ParserTransformOptions = Partial<{
  lossless: boolean
  updateSelector: boolean
}>

// normalizeTransformOptions 确保 parser 在修改选择器时保持必要的副作用
export function normalizeTransformOptions(options?: ParserTransformOptions): ParserTransformOptions {
  return {
    lossless: false,
    updateSelector: true,
    ...options,
  }
}

// mklist 将节点扩展为 "节点 + combinator + 节点副本" 的结构
export function mklist(node: Node): Node[] {
  return [
    node,
    psp.combinator({
      value: '+',
    }),
    node.clone(),
  ]
}

// composeIsPseudoAst 按需把字符串数组转换为 :is(...) 结构的 AST
export function composeIsPseudoAst(strs: string | string[]): Node[] {
  if (typeof strs === 'string') {
    return mklist(psp.tag({
      value: strs,
    }))
  }
  if (strs.length > 1) {
    return mklist(psp.pseudo({
      value: ':is',
      nodes: strs.map(str =>
        psp.tag({
          value: str,
        }),
      ),
    }))
  }
  return mklist(psp.tag({
    value: strs[0],
  }))
}

// 根据配置生成替换子代选择器的 AST，默认使用 view 标签
export function getCombinatorSelectorAst(options: IStyleHandlerOptions) {
  let childCombinatorReplaceValue: Node[] = mklist(psp.tag({ value: 'view' }))
  const { cssChildCombinatorReplaceValue } = options
  if (
    typeof cssChildCombinatorReplaceValue === 'string'
    || (Array.isArray(cssChildCombinatorReplaceValue) && cssChildCombinatorReplaceValue.length > 0)) {
    childCombinatorReplaceValue = composeIsPseudoAst(cssChildCombinatorReplaceValue)
  }
  return childCombinatorReplaceValue
}
