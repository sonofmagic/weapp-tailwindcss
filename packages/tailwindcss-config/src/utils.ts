export {
  defu,
  defuOverrideArray,
} from '@weapp-tailwindcss/shared'
