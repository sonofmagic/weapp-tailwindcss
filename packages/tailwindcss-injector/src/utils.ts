export {
  defu,
  defuOverrideArray,
  isRegexp,
  regExpTest,
} from '@weapp-tailwindcss/shared'

export {
  md5,
  removeAllExtensions as removeFileExtension,
} from '@weapp-tailwindcss/shared/node'
