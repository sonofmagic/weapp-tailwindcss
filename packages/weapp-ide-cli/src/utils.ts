import process from 'node:process'
import path from 'pathe'
import { execa } from 'execa'
import type { AliasEntry } from './types'

export async function execute(cliPath: string, argv: string[]) {
  const task = execa(cliPath, argv)

  task?.stdout?.pipe(process.stdout)

  await task
  // 调用返回码为 0 时代表正常，为 -1 时错误。
  // task.exitCode === 0
  // task.exitCode === -1
}

export function resolvePath(filePath: string) {
  if (path.isAbsolute(filePath)) {
    return filePath
  }
  else {
    return path.resolve(process.cwd(), filePath)
  }
}

function alias(argv: string[], entry: AliasEntry) {
  let findIdx = argv.indexOf(entry.find)
  // alias -p as --project
  if (findIdx > -1) {
    argv[findIdx] = entry.replacement
  }
  else {
    findIdx = argv.indexOf(entry.replacement)
  }

  if (findIdx > -1) {
    const paramIdx = findIdx + 1
    const param = argv[paramIdx]
    // 存在项目目录
    if (param && param[0] !== '-') {
      argv[paramIdx] = resolvePath(param)
    }
    else {
      argv.splice(paramIdx, 0, process.cwd())
    }
  }
  return argv
}

function pathCompat(argv: string[], option: string) {
  const findIdx = argv.indexOf(option)

  if (findIdx > -1) {
    const paramIdx = findIdx + 1
    const param = argv[paramIdx]
    // 存在项目目录
    if (param && param[0] !== '-') {
      argv[paramIdx] = resolvePath(param)
    }
    else {
      argv.splice(paramIdx, 0, process.cwd())
    }
  }
  return argv
}

export function createAlias(entry: AliasEntry) {
  return function (argv: string[]) {
    return alias(argv, entry)
  }
}

export function createPathCompat(option: string) {
  return function (argv: string[]) {
    return pathCompat(argv, option)
  }
}
