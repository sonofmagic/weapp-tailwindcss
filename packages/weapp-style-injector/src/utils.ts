import type { PerFileImportResolver } from './core'

export function toArray<T>(value: T | T[] | null | undefined): Array<NonNullable<T>> {
  if (value == null) {
    return []
  }
  return (Array.isArray(value) ? value : [value]) as Array<NonNullable<T>>
}

export function mergePerFileResolvers(
  resolvers: Array<PerFileImportResolver | null | undefined>,
): PerFileImportResolver | undefined {
  const activeResolvers = resolvers.filter((resolver): resolver is PerFileImportResolver => typeof resolver === 'function')

  if (activeResolvers.length === 0) {
    return undefined
  }

  return (fileName: string) => activeResolvers.flatMap(resolver => toArray(resolver(fileName)))
}

export function ensurePosix(value: string): string {
  return value.replace(/\\/g, '/')
}

export function normalizeRoot(root: string): string {
  const trimmed = root.trim().replace(/^[./\\]+/, '').replace(/\\+/g, '/')
  return trimmed.endsWith('/') ? trimmed.slice(0, -1) : trimmed
}

export function normalizeRelativeImport(target: string): string {
  if (target.startsWith('.') || target.startsWith('/')) {
    return target
  }
  return `./${target}`
}
