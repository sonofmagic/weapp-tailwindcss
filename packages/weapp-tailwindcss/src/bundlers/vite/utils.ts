import type { ExistingRawSourceMap } from 'rollup'
import path from 'node:path'
import process from 'node:process'

export function slash(p: string): string {
  return p.replace(/\\/g, '/')
}

export const isWindows = process.platform === 'win32'

const cssLangs = `\\.(css|less|sass|scss|styl|stylus|pcss|postcss)($|\\?)`
export const cssLangRE = new RegExp(cssLangs)
export function isCSSRequest(request: string): boolean {
  return cssLangRE.test(request)
}

export function normalizePath(id: string): string {
  return path.posix.normalize(isWindows ? slash(id) : id)
}

const postfixRE = /[?#].*$/
export function cleanUrl(url: string): string {
  return url.replace(postfixRE, '')
}

export async function formatPostcssSourceMap(
  rawMap: ExistingRawSourceMap,
  file: string,
): Promise<ExistingRawSourceMap> {
  const inputFileDir = path.dirname(file)

  const sources = rawMap.sources.map((source) => {
    const cleanSource = cleanUrl(decodeURIComponent(source))

    // postcss virtual files
    if (cleanSource[0] === '<' && cleanSource.endsWith('>')) {
      return `\0${cleanSource}`
    }

    return normalizePath(path.resolve(inputFileDir, cleanSource))
  })

  return {
    file,
    mappings: rawMap.mappings,
    names: rawMap.names,
    sources,
    sourcesContent: rawMap.sourcesContent,
    version: rawMap.version,
  }
}
