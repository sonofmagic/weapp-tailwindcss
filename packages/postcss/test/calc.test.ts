import path from 'pathe'
import { createStyleHandler } from '@/handler'
import { generateCss } from './utils'

describe('calc', () => {
  it('默认的情况', async () => {
    const code = await generateCss(path.resolve(__dirname, './fixtures/issues/calc'))
    expect(code.css).toMatchSnapshot()
  })

  it('设置 cssCalc true', async () => {
    const code = await generateCss(path.resolve(__dirname, './fixtures/issues/calc'))

    const styleHandler = createStyleHandler({
      isMainChunk: true,
      cssCalc: true,
    })
    const { css } = await styleHandler(code.css)
    expect(css).toMatchSnapshot()
  })

  it('传入 cssCalc 配置', async () => {
    const code = await generateCss(path.resolve(__dirname, './fixtures/issues/calc'))

    const styleHandler = createStyleHandler({
      isMainChunk: true,
      cssCalc: {
        mediaQueries: true,
        precision: 5,
        preserve: false,
        selectors: true,
      },
    })
    const { css } = await styleHandler(code.css)
    expect(css).toMatchSnapshot()
  })

  it('同时确保 custom-properties 也开启', async () => {
    const code = await generateCss(path.resolve(__dirname, './fixtures/issues/calc'))

    const styleHandler = createStyleHandler({
      isMainChunk: true,
      cssCalc: true,
    })
    const { css } = await styleHandler(code.css)
    expect(css).toMatchSnapshot()
  })

  it('cssCalc 和 rem2rpx 同时开启', async () => {
    const code = await generateCss(path.resolve(__dirname, './fixtures/issues/calc'))

    const styleHandler = createStyleHandler({
      isMainChunk: true,
      cssCalc: true,
      rem2rpx: true,
    })
    const { css } = await styleHandler(code.css)
    expect(css).toMatchSnapshot()
  })

  it('custom-properties 设置 preserve 为 true', async () => {
    const code = await generateCss(path.resolve(__dirname, './fixtures/issues/calc'))

    const styleHandler = createStyleHandler({
      isMainChunk: true,
      cssPresetEnv: {
        features: {
          'custom-properties': {
            preserve: true,
          },
        },
      },
    })
    const { css } = await styleHandler(code.css)
    expect(css).toMatchSnapshot()
  })

  it('rpx 运算', async () => {
    const code = `.a{ width: calc(100rpx + 100rpx); height: calc(100 * 100rpx);width: calc(68rpx * 4);height: calc(4 * 17rpx);width: calc(68 * 4rpx);height: calc(4rpx * 17);}`

    const styleHandler = createStyleHandler({
      isMainChunk: true,
      cssCalc: true,
      // cssPresetEnv: {
      //   features: {
      //     'custom-properties': {
      //       preserve: true,
      //     },
      //   },
      // },
    })
    const { css } = await styleHandler(code)
    expect(css).toMatchSnapshot()
  })

  it('rpx + root 变量运算', async () => {
    const code = `:root{--ch:2}; .a{ width: calc(var(--ch) * 1rpx);}`

    const styleHandler = createStyleHandler({
      isMainChunk: true,
      cssCalc: true,
      cssPresetEnv: {
        features: {
          'custom-properties': {
            preserve: true,
          },
        },
      },
    })
    const { css } = await styleHandler(code)
    expect(css).toMatchSnapshot()
  })

  it('rpx + root 变量运算去除', async () => {
    const code = `:root{--ch:2}; .a{ width: calc(var(--ch) * 1rpx);}`

    const styleHandler = createStyleHandler({
      isMainChunk: true,
      cssCalc: true,
      cssPresetEnv: {
        features: {
          'custom-properties': {
            preserve: false,
          },
        },
      },
    })
    const { css } = await styleHandler(code)
    expect(css).toMatchSnapshot()
  })

  it('px 转化 + root 变量运算去除', async () => {
    const code = `:root{--ch:2}; .a{ width: calc(var(--ch) * 1px);}`

    const styleHandler = createStyleHandler({
      isMainChunk: true,
      cssCalc: true,
      px2rpx: true,
    })
    const { css } = await styleHandler(code)
    expect(css).toMatchSnapshot()
  })
})
