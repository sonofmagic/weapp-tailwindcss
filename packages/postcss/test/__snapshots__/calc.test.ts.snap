// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`calc > cssCalc 和 rem2rpx 同时开启 1`] = `
"page,:host {
  --spacing: 8rpx;
}
view,text,:after,:before,::backdrop {
  box-sizing: border-box;
  border: 0 solid;
  margin: 0;
  padding: 0;
}
.w-1 {
  width: 8rpx;
}"
`;

exports[`calc > custom-properties 设置 preserve 为 true 1`] = `
"page,:host {
  --spacing: .25rem;
}
view,text,:after,:before,::backdrop {
  box-sizing: border-box;
  border: 0 solid;
  margin: 0;
  padding: 0;
}
.w-1 {
  width: calc(.25rem * 1);
  width: calc(var(--spacing) * 1);
}"
`;

exports[`calc > px 转化 + root 变量运算去除 1`] = `"page{--ch:2}; .a{ width: 2rpx;}"`;

exports[`calc > rpx + root 变量运算 1`] = `"page{--ch:2}; .a{ width: 2rpx; width: calc(var(--ch)*1rpx);}"`;

exports[`calc > rpx + root 变量运算去除 1`] = `"page{--ch:2}; .a{ width: 2rpx;}"`;

exports[`calc > rpx 运算 1`] = `".a{ width: 200rpx; height: 10000rpx;width: 272rpx;height: 68rpx;width: 272rpx;height: 68rpx;}"`;

exports[`calc > 传入 cssCalc 配置 1`] = `
"page,:host {
  --spacing: .25rem;
}
view,text,:after,:before,::backdrop {
  box-sizing: border-box;
  border: 0 solid;
  margin: 0;
  padding: 0;
}
.w-1 {
  width: 0.25rem;
}"
`;

exports[`calc > 同时确保 custom-properties 也开启 1`] = `
"page,:host {
  --spacing: .25rem;
}
view,text,:after,:before,::backdrop {
  box-sizing: border-box;
  border: 0 solid;
  margin: 0;
  padding: 0;
}
.w-1 {
  width: 0.25rem;
}"
`;

exports[`calc > 设置 cssCalc true 1`] = `
"page,:host {
  --spacing: .25rem;
}
view,text,:after,:before,::backdrop {
  box-sizing: border-box;
  border: 0 solid;
  margin: 0;
  padding: 0;
}
.w-1 {
  width: 0.25rem;
}"
`;

exports[`calc > 默认的情况 1`] = `
":root, :host {
  --spacing: .25rem;
}
*,:after,:before,::backdrop {
  box-sizing: border-box;
  border: 0 solid;
  margin: 0;
  padding: 0;
}
.w-1 {
  width: calc(var(--spacing) * 1);
}"
`;
